luarules/gadgets/unit_projectile_leash.lua

```diff
@@ -0,0 +1,163 @@
+function gadget:GetInfo()
+	return {
+		name = "Projectile Over-Range and Leashing",
+		desc = "Destroys projectiles if they exceed defined ranges",
+		author = "SethDGamre",
+		layer = 1,
+		enabled = true
+	}
+end
+
+if not gadgetHandler:IsSyncedCode() then return end
+
+---- unit customParams ----
+-- use weaponDef customparams.projectile_overrange_distance to destroy projectiles exceeding its limit.
+-- use weaponDef customparams.projectile_leash_range to destroy projectiles when they exceed range/projectile_overrange_distance and projectile_leash_range relative to unit position.
+
+---- optional customParams ----
+-- use weaponDef customparams.weaponDef.customParams.projectile_destruction_method = "string" to change how projectiles are destroyed.
+-- "explode" (default if undefined) detonates the projectile.
+-- "descend" moves the projectile downward until it is destroyed by collision event.
+
+--static values
+local lazyUpdateFrames = math.ceil(Game.gameSpeed / 3 * 2)
+local edgyUpdateFrames = math.ceil(Game.gameSpeed / 6)
+local forcedDescentUpdateFrames = math.ceil(Game.gameSpeed / 5)
+local minimumThresholdRange = 10 --so that starburst missiles don't trigger edgyWatch during ascent
+local compoundingMultiplier = 1.1 --compounding multiplier that influences the arc at which projectiles are forced to descend
+local descentSpeedStartingMultiplier = 0.1
+
+--functions
+local spGetUnitPosition = Spring.GetUnitPosition
+local mathSqrt = math.sqrt
+local mathMax = math.max
+local spGetProjectilePosition = Spring.GetProjectilePosition
+local spSetProjectileCollision = Spring.SetProjectileCollision
+local spGetProjectileVelocity = Spring.GetProjectileVelocity
+local spSetProjectileVelocity = Spring.SetProjectileVelocity
+
+--tables
+local defWatchTable = {}
+local lazyProjectileWatch = {}
+local edgyProjectileWatch = {}
+local forcedDescentTable = {}
+
+
+for weaponDefID, weaponDef in pairs(WeaponDefs) do
+	if weaponDef.customParams.projectile_leash_range then
```

Comment:
Several things here:

1. I think you can have this else after the other one, or even inside, that way you don't need to `defWatchTable[weaponDefID] = {}` two times since you know the other if is going to initialize it for sure, and can just initialize to `{}` one time.

2. Avoid repeating the long weaponDef.customParams by doing `local customParams = weaponDef.customParams` first

3. Avoid doing defWatchTable[weaponDefID] all the time by doing:

```LUA
-- init you table one time
local watchParams = {}
-- then assign everything
watchParams.range = weaponDef.range
watchParams.weaponRange = weaponDef.range
(...)
-- and at the very end:
defWatchTable[weaponDefID] = watchParams
```

Cleaner imo, and also slighter more efficient.

Reply:
donezo, thanks. 10% smarter